libraries/AP_HAL_Linux/RCOutput_Bebop.cpp

```diff
@@ -0,0 +1,440 @@
+
+    if(!_i2c_sem->take(10))
+        return;
+
+    hal.i2c1->write(BEBOP_BLDC_I2C_ADDR, 1, &data);
+
+    _i2c_sem->give();
+}
+
+void LinuxRCOutput_Bebop::_clear_error()
+{
+    uint8_t data = BEBOP_BLDC_CLEAR_ERROR;
+
+    if(!_i2c_sem->take(10))
+        return;
+
+    hal.i2c1->write(BEBOP_BLDC_I2C_ADDR, 1, &data);
+
+    _i2c_sem->give();
+}
+
+void LinuxRCOutput_Bebop::_play_sound(uint8_t sound)
+{
+    if(!_i2c_sem->take(10))
+        return;
+
+    hal.i2c1->writeRegister(BEBOP_BLDC_I2C_ADDR, BEBOP_BLDC_PLAY_SOUND, sound);
+
+    _i2c_sem->give();
+}
+
+uint16_t LinuxRCOutput_Bebop::_period_us_to_rpm(uint16_t period_us)
+{
+    float period_us_fl = period_us;
+
+    float rpm_fl = (period_us_fl - _min_pwm)/(_max_pwm - _min_pwm) *
+                    (BEBOP_BLDC_MAX_RPM - BEBOP_BLDC_MIN_RPM) + BEBOP_BLDC_MIN_RPM;
+
+    return (uint16_t)rpm_fl;
+}
+
+void LinuxRCOutput_Bebop::init(void* dummy)
+{
+    int ret=0;
+    struct sched_param param = { .sched_priority = RCOUT_BEBOP_RTPRIO };
+    pthread_attr_t attr;
+
+    _i2c_sem = hal.i2c1->get_semaphore();
+    if (_i2c_sem == NULL) {
+        hal.scheduler->panic(PSTR("RCOutput_Bebop: can't get i2c sem"));
+        return; /* never reached */
+    }
+
+    /* Initialize thread, cond, and mutex */
+    ret = pthread_mutex_init(&_mutex, NULL);
+    if(ret != 0) {
+        perror("RCout_Bebop: failed to init mutex\n");
+        goto err_mutex;
+    }
+
+    pthread_mutex_lock(&_mutex);
+
+    ret = pthread_cond_init(&_cond, NULL);
+
+    if(ret != 0) {
+        perror("RCout_Bebop: failed to init cond\n");
+        goto err_mutex;
+    }
+
+    ret = pthread_attr_init(&attr);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to init attr\n");
+        goto err_mutex;
+    }
+    pthread_attr_setinheritsched(&attr, PTHREAD_EXPLICIT_SCHED);
+    pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
+    pthread_attr_setschedparam(&attr, &param);
+    ret = pthread_create(&_thread, &attr, _control_thread, this);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to create thread\n");
+        goto err_mutex;
+    }
+
+    _clear_error();
+
+    /* Set an initial dummy frequency */
+    _frequency = 50;
+    return;
+
+err_mutex:
+    pthread_mutex_unlock(&_mutex);
+    return;
+}
+
+void LinuxRCOutput_Bebop::set_freq(uint32_t chmask, uint16_t freq_hz)
+{
+    _frequency = freq_hz;
+}
+
+uint16_t LinuxRCOutput_Bebop::get_freq(uint8_t ch)
+{
+    return _frequency;
+}
+
+void LinuxRCOutput_Bebop::enable_ch(uint8_t ch)
+{
+}
+
+void LinuxRCOutput_Bebop::disable_ch(uint8_t ch)
+{
+    _stop_prop();
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t period_us)
+{
+
+    pthread_mutex_lock(&_mutex);
+    _period_us[ch] = period_us;
+    pthread_cond_signal(&_cond);
+    pthread_mutex_unlock(&_mutex);
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t* period_us, uint8_t len)
+{
+    for (int i = 0; i < len; i++)
+        write(ch + i, period_us[i]);
```

Comment:
This suffers the same problem as the RCOutput_Navio... we write the whole vector of PWM N times, in which N is the number of motors. There's a ***hack*** here that I'm using: https://github.com/lucasdemarchi/ardupilot/commits/wip/rcoutput_navio-dirty. I'm going to rebase/clean this branch soon and send a pr, then I think I can apply it here, too.